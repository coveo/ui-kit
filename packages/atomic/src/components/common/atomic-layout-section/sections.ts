export type Section =
  | 'search'
  | 'facets'
  | 'main'
  | 'status'
  | 'results'
  | 'horizontal-facets'
  | 'products'
  | 'pagination';

export function findSection(element: HTMLElement, section: Section) {
  return element.querySelector(
    sectionSelector(section)
  ) as HTMLAtomicLayoutSectionElement | null;
}

export function sectionSelector(section: Section) {
  return `atomic-layout-section[section="${section}"]`;
}
