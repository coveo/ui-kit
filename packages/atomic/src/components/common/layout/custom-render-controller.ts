import type {LitElement, ReactiveController, ReactiveControllerHost} from 'lit';
import type {ItemRenderingFunction} from '@/src/components/common/item-list/item-list-common';
import type {AnyItem} from '@/src/components/common/item-list/unfolded-item';

export interface CustomRenderHost extends ReactiveControllerHost {
  shadowRoot?: ShadowRoot | null;
}

export interface CustomRenderOptions {
  renderingFunction: () => ItemRenderingFunction<AnyItem> | undefined;
  itemData: () => AnyItem | undefined;
  rootElementRef: () => HTMLElement | undefined;
  linkContainerRef?: () => HTMLElement | undefined;
  onRenderComplete?: (element: HTMLElement, output: string) => void;
}

/**
 * A reactive controller that manages custom rendering function execution for item components.
 */
export class CustomRenderController implements ReactiveController {
  private options: Required<CustomRenderOptions>;
  private hasExecutedRenderFunction = false;

  constructor(
    host: CustomRenderHost & LitElement,
    options: CustomRenderOptions
  ) {
    this.options = {
      linkContainerRef: () => undefined,
      onRenderComplete: () => {},
      ...options,
    };
    host.addController(this);
  }

  hostConnected(): void {
    this.resetRenderState();
  }

  hostUpdated(): void {
    this.executeRenderFunction();
  }

  private resetRenderState(): void {
    this.hasExecutedRenderFunction = false;
  }

  private executeRenderFunction(): void {
    const renderingFunction = this.options.renderingFunction();
    const itemData = this.options.itemData();
    const rootElementRef = this.options.rootElementRef();
    if (
      !renderingFunction ||
      !itemData ||
      !rootElementRef ||
      this.hasExecutedRenderFunction
    ) {
      return;
    }
    const customRenderOutput = renderingFunction(
      itemData,
      rootElementRef,
      this.options.linkContainerRef()
    );
    this.options.onRenderComplete(rootElementRef, customRenderOutput);
    this.hasExecutedRenderFunction = true;
  }
}
