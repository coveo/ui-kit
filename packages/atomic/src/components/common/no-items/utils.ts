import {i18n} from 'i18next';

export const getSummary = (
  i18n: i18n,
  query: string,
  hasResults: boolean,
  i18nKey: 'no-products' | 'no-results'
) => {
  if (hasResults) {
    return '';
  }

  const labelFor = `${i18nKey}-for`;

  return query
    ? i18n.t(labelFor, {
        interpolation: {escapeValue: false},
        query,
      })
    : i18n.t(i18nKey);
};
