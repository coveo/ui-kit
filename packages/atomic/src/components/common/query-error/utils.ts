import {i18n} from 'i18next';
import {KnownErrorType} from './known-error-types';

export const getErrorTitleFromErrorType = (
  i18n: i18n,
  organizationId: string,
  errorType?: string
) => {
  switch (errorType as KnownErrorType) {
    case 'Disconnected':
      return i18n.t('disconnected');

    case 'NoEndpointsException':
      return i18n.t('no-endpoints', {org: organizationId});
    case 'InvalidTokenException':
      return i18n.t('cannot-access', {org: organizationId});
    case 'OrganizationIsPausedException':
      return i18n.t('organization-is-paused', {org: organizationId});
    default:
      return i18n.t('something-went-wrong');
  }
};

export const getErrorDescriptionFromErrorType = (
  i18n: i18n,
  organizationId: string,
  url: string,
  errorType?: string
) => {
  switch (errorType as KnownErrorType) {
    case 'Disconnected':
      return i18n.t('check-your-connection', {url});
    case 'NoEndpointsException':
      return i18n.t('add-sources');
    case 'InvalidTokenException':
      return i18n.t('invalid-token');
    case 'OrganizationIsPausedException':
      return i18n.t('organization-will-resume', {
        org: organizationId,
      });
    default:
      return i18n.t('if-problem-persists');
  }
};

export const getAriaMessageFromErrorType = (
  i18n: i18n,
  organizationId: string,
  platformUrl: string,
  errorType?: string
) => {
  return `${getErrorTitleFromErrorType(i18n, organizationId, errorType)} ${getErrorDescriptionFromErrorType(i18n, organizationId, platformUrl, errorType)}`;
};
