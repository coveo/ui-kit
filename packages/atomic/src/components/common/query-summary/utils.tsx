import {FunctionalComponent, h} from '@stencil/core';
import {i18n} from 'i18next';

interface i18nKeyProps {
  first: number;
  last: number;
  total: number;
  query: string;
  isLoading: boolean;
  i18n: i18n;
}

const WrapHighlight: FunctionalComponent<{part?: string}> = (
  props,
  children
) => {
  return (
    <span
      class="font-bold"
      part={`highlight${props.part ? ` ${props.part}` : ''}`}
    >
      {children}
    </span>
  );
};

export const getQuerySummaryI18nParameters = ({
  first,
  last,
  query,
  total,
  isLoading,
  i18n,
}: i18nKeyProps) => {
  const i18nKey =
    query !== '' ? 'showing-results-of-with-query' : 'showing-results-of';

  const params = {
    first: first.toLocaleString(),
    last: last.toLocaleString(),
    total: total.toLocaleString(),
    query: query,
    count: last,
  };

  const highlights = {
    first: <WrapHighlight>{params.first}</WrapHighlight>,
    last: <WrapHighlight>{params.last}</WrapHighlight>,
    total: <WrapHighlight>{params.total}</WrapHighlight>,
    query: <WrapHighlight part="query">{params.query}</WrapHighlight>,
  };

  const ariaLiveMessage = isLoading
    ? i18n.t('loading-results')
    : i18n.t(i18nKey, params);

  return {i18nKey, highlights, ariaLiveMessage};
};
