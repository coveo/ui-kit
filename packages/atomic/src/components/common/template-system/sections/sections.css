@import "./section-bottom-metadata.css";
@import "./section-emphasized.css";
@reference '../../../../utils/tailwind.global.tw.css';
