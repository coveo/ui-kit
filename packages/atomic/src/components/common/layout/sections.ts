import {isElementNode} from '../../../utils/utils';

const resultSectionTags = new Set([
  'atomic-result-section-visual',
  'atomic-result-section-badges',
  'atomic-result-section-actions',
  'atomic-result-section-title',
  'atomic-result-section-title-metadata',
  'atomic-result-section-emphasized',
  'atomic-result-section-excerpt',
  'atomic-result-section-bottom-metadata',
  'atomic-result-section-children',
] as const);

const productSectionTags = new Set([
  'atomic-product-section-visual',
  'atomic-product-section-badges',
  'atomic-product-section-actions',
  'atomic-product-section-name',
  'atomic-product-section-metadata',
  'atomic-product-section-emphasized',
  'atomic-product-section-description',
  'atomic-product-section-bottom-metadata',
] as const);

const allTags = new Set([...resultSectionTags, ...productSectionTags]);

type SetValueType<T> = T extends Set<infer U> ? U : never;

export type ItemSectionTagName = SetValueType<typeof allTags>;

export function isResultSectionNode(element: Node) {
  if (!isElementNode(element)) {
    return false;
  }
  return allTags.has(element.tagName.toLowerCase() as ItemSectionTagName);
}

export function containsSections(content: string | NodeList | HTMLCollection) {
  if (typeof content === 'string') {
    return Array.from(resultSectionTags.values()).some((resultSectionTag) =>
      content.includes(resultSectionTag)
    );
  }
  return Array.from(content).some((child) => isResultSectionNode(child));
}
