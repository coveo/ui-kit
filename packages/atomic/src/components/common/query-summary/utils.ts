import {FunctionalComponent} from '@/src/utils/functional-component-utils';
import {i18n} from 'i18next';
import {html} from 'lit';

interface i18nKeyProps {
  first: number;
  last: number;
  total: number;
  query: string;
  isLoading: boolean;
  i18n: i18n;
}

interface QuerySummaryLabels {
  loadingStatus: string;
  itemsForQuery: string;
  allItems: string;
}

const renderWrapHighlight: FunctionalComponent<{
  part?: string;
  content: string;
}> = ({props}) => {
  const part = `highlight${props.part ? ` ${props.part}` : ''}`;
  return html` <span class="front-bold" part="${part}">${props.content}</span>`;
};

export const getQuerySummaryI18nParameters = (props: i18nKeyProps) => {
  return getQuerySummaryData(props, {
    loadingStatus: 'loading-results',
    itemsForQuery: 'showing-results-of-with-query',
    allItems: 'showing-results-of',
  });
};

export const getProductQuerySummaryI18nParameters = (props: i18nKeyProps) => {
  return getQuerySummaryData(props, {
    loadingStatus: 'loading-products',
    itemsForQuery: 'showing-products-of-with-query',
    allItems: 'showing-products-of',
  });
};

const getQuerySummaryData = (
  {first, last, query, total, isLoading, i18n}: i18nKeyProps,
  {allItems, itemsForQuery, loadingStatus}: QuerySummaryLabels
) => {
  const i18nKey = query === '' ? allItems : itemsForQuery;

  const locale = i18n.language;
  const params = {
    first: first.toLocaleString(locale),
    last: last.toLocaleString(locale),
    total: total.toLocaleString(locale),
    query: query,
    count: last,
  };

  const highlights = {
    first: renderWrapHighlight({props: {content: params.first}}),
    last: renderWrapHighlight({props: {content: params.last}}),
    total: renderWrapHighlight({props: {content: params.total}}),
    query: renderWrapHighlight({props: {part: 'query', content: params.query}}),
  };

  const ariaLiveMessage = isLoading
    ? i18n.t(loadingStatus)
    : i18n.t(i18nKey, params);

  return {i18nKey, highlights, ariaLiveMessage};
};
