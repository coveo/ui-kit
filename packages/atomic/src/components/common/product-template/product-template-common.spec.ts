import {
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
  MockInstance,
  vi,
} from 'vitest';
import {isElementNode, isVisualNode} from '@/src/utils/utils';
import {isResultSectionNode} from '../layout/sections';
import {
  getTemplateNodeType,
  makeDefinedConditions,
  makeMatchConditions,
} from './product-template-common';

vi.mock('../layout/sections', {spy: true});
vi.mock('@/src/utils/utils', {spy: true});
vi.mock('@coveo/headless/commerce');

describe('utils', () => {
  describe('makeMatchConditions', () => {
    let consoleErrorSpy: MockInstance;

    beforeEach(() => {
      consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {});
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('should log a warning and return an always-false callback when conditions are conflicting', () => {
      const conditions = makeMatchConditions(
        {field: ['value']},
        {field: ['value']}
      );
      expect(conditions).toHaveLength(1);
      expect(conditions[0].toString()).toMatchSnapshot();
      expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
      expect(consoleErrorSpy.mock.calls[0]).toMatchSnapshot();
    });

    it('should return a mustMatch and a mustNotMatch condition when no conflict', () => {
      const conditions = makeMatchConditions(
        {field: ['value1']},
        {field: ['value2']}
      );
      expect(conditions.length).toBe(2);
      expect(typeof conditions[0]).toBe('function');
      expect(typeof conditions[1]).toBe('function');
    });

    it('should return only mustNotMatch conditions if mustMatch is empty', () => {
      const conditions = makeMatchConditions({}, {field: ['value2']});
      expect(conditions.length).toBe(1);
      expect(typeof conditions[0]).toBe('function');
    });

    it('should return only mustMatch conditions if mustNotMatch is empty', () => {
      const conditions = makeMatchConditions({field: ['value1']}, {});
      expect(conditions.length).toBe(1);
      expect(typeof conditions[0]).toBe('function');
    });
  });

  describe('makeDefinedConditions', () => {
    it('should return a condition for ifDefined', () => {
      const conditions = makeDefinedConditions('foo,bar', undefined);
      expect(conditions.length).toBe(1);
      expect(typeof conditions[0]).toBe('function');
    });

    it('should return a condition for ifNotDefined', () => {
      const conditions = makeDefinedConditions(undefined, 'baz');
      expect(conditions.length).toBe(1);
      expect(typeof conditions[0]).toBe('function');
    });

    it('should return two conditions for both ifDefined and ifNotDefined', () => {
      const conditions = makeDefinedConditions('foo', 'bar');
      expect(conditions.length).toBe(2);
      expect(typeof conditions[0]).toBe('function');
      expect(typeof conditions[1]).toBe('function');
    });

    it('should return an empty array if neither is provided', () => {
      const conditions = makeDefinedConditions();
      expect(conditions.length).toBe(0);
    });
  });

  describe('getTemplateNodeType', () => {
    it('should return "section" for a node with isResultSectionNode true', () => {
      const node: Node = document.createElement('anything');
      vi.mocked(isResultSectionNode).mockReturnValueOnce(true);
      expect(getTemplateNodeType(node)).toBe('section');
    });

    it('should return "metadata" for a node that is not visual', () => {
      const node: Node = document.createElement('anything');
      vi.mocked(isResultSectionNode).mockReturnValueOnce(false);
      vi.mocked(isVisualNode).mockReturnValueOnce(false); // dsa
      expect(getTemplateNodeType(node)).toBe('metadata');
    });

    it('should return "table-column-definition" for a table element', () => {
      const node: Node = document.createElement('atomic-table-element');
      vi.mocked(isResultSectionNode).mockReturnValueOnce(false);
      vi.mocked(isVisualNode).mockReturnValueOnce(true);
      vi.mocked(isElementNode).mockReturnValueOnce(true);
      expect(getTemplateNodeType(node)).toBe('table-column-definition');
    });

    it('should return "other" for any other node', () => {
      const node: Node = document.createElement('anything');
      vi.mocked(isResultSectionNode).mockReturnValueOnce(false);
      vi.mocked(isVisualNode).mockReturnValueOnce(true);
      vi.mocked(isElementNode).mockReturnValueOnce(false);
      expect(getTemplateNodeType(node)).toBe('other');
    });
  });
});
