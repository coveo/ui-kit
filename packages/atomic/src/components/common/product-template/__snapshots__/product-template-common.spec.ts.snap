// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > makeMatchConditions > should log a warning and return an always-false callback when conditions are conflicting 1`] = `"() => false"`;

exports[`utils > makeMatchConditions > should log a warning and return an always-false callback when conditions are conflicting 2`] = `
[
  "Conflicting match conditions for field field, the template will be ignored.",
  Set {
    "value",
  },
]
`;
